google/cloud/bigtable/examples/stock_backtesting/backtesting_main.cc

```diff
@@ -0,0 +1,259 @@
+              << "<start_date> <end_date> <project_id> <instance_id> <table_id>"
+              << std::endl;
+    return 1;
+  }
+
+  // Prerequisite check.
+  std::string const strategy_filepath = argv[1];
+  std::string ticker = argv[2];
+  absl::AsciiStrToUpper(&ticker);
+  std::string const start_date_str = argv[3];
+  std::string const end_date_str = argv[4];
+  std::string const project_id = argv[5];
+  std::string const instance_id = argv[6];
+  std::string const table_id = argv[7];
+  if (strategy_filepath.empty() || ticker.empty() || start_date_str.empty() ||
+      end_date_str.empty() || project_id.empty() || instance_id.empty() ||
+      table_id.empty()) {
+    std::cerr << "Please specify necessary parameters." << std::endl;
+    return 1;
+  }
+
+  absl::CivilDay start_date, end_date, min_start_date, max_end_date;
+  if (!absl::ParseCivilTime(start_date_str, &start_date) ||
+      !absl::ParseCivilTime(end_date_str, &end_date)) {
+    std::cerr << "Can't parse the dates: " << start_date_str << " and "
+              << end_date_str << std::endl;
+    return 1;
+  }
+  if (!absl::ParseCivilTime(kMinStartDate, &min_start_date) ||
+      !absl::ParseCivilTime(kMaxEndDate, &max_end_date)) {
+    std::cerr << "Can't parse the backtesting window." << std::endl;
+    return 1;
+  }
+  if (start_date < min_start_date || end_date > max_end_date ||
+      start_date >= end_date) {
+    std::cerr << "Backtesting only supports time window "
+              << "[2016-10-31, 2021-10-29).";
+    return 1;
+  }
+
+  // Read the input strategy file.
+  std::fstream input_file(strategy_filepath);
+  if (!input_file.is_open()) {
+    std::cerr << "Error in opening file: " << strategy_filepath << std::endl;
+    return 1;
+  }
+  std::string input_str(std::istreambuf_iterator<char>{input_file}, {});
+  cbt::examples::Strategy strategy;
+  if (!google::protobuf::TextFormat::ParseFromString(input_str, &strategy)) {
+    std::cerr << "Can't parse the input strategy." << std::endl;
+    return 1;
+  }
+
+  // Prepare the table.
+  cbt::Table table(cbt::MakeDataClient(project_id, instance_id), table_id);
+
+  auto const row_keys = PrepareRowKeys(ticker, start_date, end_date);
+  auto const filter = PrepareFilter(strategy, start_date, end_date);
+
+  // A map from column_name/signal to {date, price} map;
+  // Note the inner map is sorted by date.
+  std::unordered_map<std::string, std::map<absl::CivilDay, double>> signal_map;
+
+  for (auto const& row_key : row_keys) {
+    auto result = table.ReadRow(row_key, filter).value();
+    if (!result.first) {
+      std::cerr << "Error in reading row key: " << row_key << "; continue."
+                << std::endl;
+      continue;
+    }
+
+    auto const& cells = result.second.cells();
+    for (auto const& cell : cells) {
+      absl::Duration const ts_duration = absl::FromChrono(cell.timestamp());
+      absl::Time const ts_time =
+          absl::time_internal::FromUnixDuration(ts_duration);
```

Comment:
You are right that we don't have to have them in this example, e.g. if I change the signature of `signal_map` to use `std::milliseconds` as the inner key. But I recall that I struggled with the conversion between `absl::time_internal` and `std::chrono` when I was working on this example; it might be good to keep them as an example to show how the conversion could be done. Therefore I decided to keep it.
I added a comment to clarify this.

Reply:
There is no conversion to/from `absl::time_internal` because it is not a class ... it is a namespace that you should not need to use.

In this case, for example, to convert from a "Unix" `absl::Duration` to an `absl::Time`, simply add `absl::UnixEpoch()`.

Reply:
Sorry Bradley, I didn't get the point.
The timestamp we got from cell is a std::chrono, and my `signal_map` key is a absl::CivilDay, to convert std::chrono to absl::CivilDay, I did
1. Convert std::chrono to absl::Duration
2. Convert absl::Duration to absl::Time
3. Convert absl::Time to absl::CivilDay

What do you suggest instead?

Reply:
Sorry for not being clearer.  The suggestion for "Convert absl::Duration to absl::Time" is to change
```
absl::Time const ts_time = absl::time_internal::FromUnixDuration(ts_duration);
```
which calls an _internal-only_ function, to
```
absl::Time const ts_time = absl::UnixEpoch() + ts_duration;
```
which only uses the advertised API.

Reply:
Thanks for the comments! Done.